### ISSUE TITLE
e2e test case for udp backend is failing on s390x

### ISSUE THREAD

**Issue by salamaniibm** [coreos/flannel#655]
I am working on s390x architecture.
The e2e-test is failing only for udp backend. where as the vxlan and  host-gw are successfull.

- make e2e-test ARCH=s390x BACKEND=udp
```
=== BACKEND: udp ===============================================
{ "Network": "10.10.0.0/16", "Backend": { "Type": "udp" } }
flannel config written
flannel-e2e-test-flannel1
4281429babc9b34b3a9873268490ed3dbc11c1147198431637105198d5cf4a69
flannel-e2e-test-flannel2
24e372e3c8b6d95bcc715fb1263c8d0f4d5a024691a07d280dbab05e8d1a676a
flannels running
PING 10.10.66.1 (10.10.66.1): 56 data bytes

--- 10.10.66.1 ping statistics ---
5 packets transmitted, 0 packets received, 100% packet loss
Test for backend=udp: exit=1
------ flannel server (one being pinged) log -------
I0327 09:42:04.676230       1 main.go:119] Installing signal handlers
I0327 09:42:04.680609       1 manager.go:136] Determining IP address of default interface
I0327 09:42:04.680855       1 manager.go:149] Using interface with name eth0 and address 172.17.0.3
I0327 09:42:04.680867       1 manager.go:166] Defaulting external address to interface address (172.17.0.3)
I0327 09:42:04.693985       1 local_manager.go:179] Picking subnet in range 10.10.1.0 ... 10.10.255.0
I0327 09:42:04.699945       1 manager.go:250] Lease acquired: 10.10.66.0/24
I0327 09:42:04.736589       1 network.go:99] Watching for new subnet leases
I0327 09:42:07.662605       1 network.go:191] Subnet added: 10.10.73.0/24

------ flannel client (one doing the ping) log -------
I0327 09:42:07.636985       1 main.go:119] Installing signal handlers
I0327 09:42:07.645022       1 manager.go:136] Determining IP address of default interface
I0327 09:42:07.645356       1 manager.go:149] Using interface with name eth0 and address 172.17.0.4
I0327 09:42:07.645379       1 manager.go:166] Defaulting external address to interface address (172.17.0.4)
I0327 09:42:07.655594       1 local_manager.go:179] Picking subnet in range 10.10.1.0 ... 10.10.255.0
I0327 09:42:07.662961       1 manager.go:250] Lease acquired: 10.10.73.0/24
I0327 09:42:07.708617       1 network.go:99] Watching for new subnet leases
I0327 09:42:07.717889       1 network.go:191] Subnet added: 10.10.66.0/24

------ etcd log -------
"docker logs" requires exactly 1 argument(s).
See 'docker logs --help'.

Usage:  docker logs [OPTIONS] CONTAINER

Fetch the logs of a container

TEST(S) FAILED
Makefile:55: recipe for target 'e2e-test' failed
make: *** [e2e-test] Error 1
```

- Commented the functional test case to avoid deleting of flannel and etcd containers after its execution.
The status of flannel-e2e-test-flannel1 container:
docker exec -it --privileged flannel-e2e-test-flannel1 /bin/sh
```
/ # ip address
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: flannel0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1472 qdisc pfifo_fast qlen 500
    link/[65534]
    inet 10.10.66.0/16 scope global flannel0
       valid_lft forever preferred_lft forever
3: dummy0: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue qlen 1000
    link/ether 86:5c:b6:bb:1b:f9 brd ff:ff:ff:ff:ff:ff
    inet 10.10.66.1/24 scope global dummy0
       valid_lft forever preferred_lft forever
    inet6 fe80::845c:b6ff:febb:1bf9/64 scope link
       valid_lft forever preferred_lft forever
140: eth0@if141: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue
    link/ether 02:42:ac:11:00:03 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.3/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:3/64 scope link
       valid_lft forever preferred_lft forever

#ping to flannel subnet within container
/ # ping -c 2 10.10.66.0
PING 10.10.66.0 (10.10.66.0): 56 data bytes
64 bytes from 10.10.66.0: seq=0 ttl=64 time=0.069 ms
64 bytes from 10.10.66.0: seq=1 ttl=64 time=0.116 ms

--- 10.10.66.0 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.069/0.092/0.116 ms

/ # ip route
default via 172.17.0.1 dev eth0
10.10.0.0/16 dev flannel0  src 10.10.66.0
10.10.66.0/24 dev dummy0  src 10.10.66.1
172.17.0.0/16 dev eth0  src 172.17.0.3
```
- The status of flannel-e2e-test-flannel2 container:
docker exec -it --privileged flannel-e2e-test-flannel2 /bin/sh
```
/ # ip address
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: flannel0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1472 qdisc pfifo_fast qlen 500
    link/[65534]
    inet 10.10.73.0/16 scope global flannel0
       valid_lft forever preferred_lft forever
142: eth0@if143: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue
    link/ether 02:42:ac:11:00:04 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.4/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:4/64 scope link
       valid_lft forever preferred_lft forever
       
# ping flannel-e2e-test-flannel1 dummy0 interface   
/ # ping -c 5 10.10.66.1
PING 10.10.66.1 (10.10.66.1): 56 data bytes

--- 10.10.66.1 ping statistics ---
5 packets transmitted, 0 packets received, 100% packet loss

# ping flannel-e2e-test-flannel1 flannel subnet
/ # ping -c 5 10.10.66.0
PING 10.10.66.0 (10.10.66.0): 56 data bytes

--- 10.10.66.0 ping statistics ---
5 packets transmitted, 0 packets received, 100% packet loss

/ # iproute
default via 172.17.0.1 dev eth0
10.10.0.0/16 dev flannel0  src 10.10.73.0
172.17.0.0/16 dev eth0  src 172.17.0.4
```

**Comment by tomdee** [comment]
Interesting. UDP is the only backend with C code. Maybe something in there isn't s390x compatible?

**Comment by guirish** [comment]
As s390x uses Big endian format, I saw all routing table IP's in reverse order due to changes in byte alignment in C code. I am able to pass the test cases by removing networkOrder function calls in cproxy.go.

I am currently looking for a fix either by conditionally checking endianness.

**Comment by tomdee** [comment]
Sounds good, thanks for tracking it down.

**Comment by tomdee** [comment]
Closed by #703
